### PR TITLE
[TypeDeclaration] Add ReturnTypeFromStrictFluentReturnRector

### LINKED ISSUE #8067: Infer return type from `$this`
# Feature Request

I would expect one of the rules out of `SetList::TYPE_DECLARATION` would infer a return type when a method returns just `$this`.

https://getrector.com/demo/84754cac-d24e-4286-a8ee-370c4a87d5a7
 
is this missing from one of the rectors, or is this something achieved by a different rector? I couldn't find one, even in other sets

### PR DESCRIPTION
**Pull request by samsonasik**
@TomasVotruba @staabm let's add new rule `ReturnTypeFromStrictFluentReturnRector` to add return type on `return $this;` usage.

Closes https://github.com/rectorphp/rector/issues/8067

### CODE REVIEW

File: rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/non_final_class.php.inc [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+?>
+-----
+<?php
+
+namespace Rector\Tests\TypeDeclaration\Rector\ClassMethod\ReturnTypeFromStrictFluentReturnRector\Fixture;
+
+class NonFinalClass
+{
+    public function test(): static
```
Comment (samsonasik) [inline rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/non_final_class.php.inc:21]: @staabm when return `$this`, return `static` or return `self` is actually same, see

- https://3v4l.org/U6gfm#v7.0.33 on PHP 7.0
- https://3v4l.org/B61Io#v8.0.29 on PHP 8.0 (support return static)

so the fallback to return `self` is fine on php 7.x

Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/non_final_class.php.inc:21]: I added test for php 7.x add return `self` instead of `static` https://github.com/rectorphp/rector-src/pull/4890/commits/5e5ba5e680d10103d0e993fd80c4035c4af7b176

File: rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/multi_return_this.php.inc [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+final class MultiReturnThis
+{
+    private $foo = 'bar';
+
+    public function test(): self
+    {
+        if (rand(0, 1)) {
+            $this->foo = 'foo';
+            return $this;
```
Comment (staabm) [inline rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/multi_return_this.php.inc:34]: Another test with a `return $this` and `return null` case would be great to support nullable

File: rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/inner_classmethod.php.inc [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+namespace Rector\Tests\TypeDeclaration\Rector\ClassMethod\ReturnTypeFromStrictFluentReturnRector\Fixture;
+
+final class InnerClassMethod
+{
+    public function test()
+    {
+        new class {
+            public function run()
+            {
+                return $this;
```
Comment (staabm) [inline rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/inner_classmethod.php.inc:12]: Do we need support for ` return new self()` or `return new static();` ?

File: rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/multi_return_this.php.inc (hunk 2 above)
Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/multi_return_this.php.inc:34]: that's already covered on `ReturnUnionTypeRector` https://getrector.com/demo/cda3b293-ac7d-4432-83f5-f23cc833a4c1 while return `?static` on final class is not needed, that should be `?self`, but that will need separate PR for improvement and consideration.

File: rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/inner_classmethod.php.inc (hunk 3 above)
Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/inner_classmethod.php.inc:12]: I think that need separate PR for easier to review ;)

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/multi_return_this.php.inc (hunk 2 above)
Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictFluentReturnRector/Fixture/multi_return_this.php.inc:34]: I created PR:

- https://github.com/rectorphp/rector-src/pull/4891

for nullable `$this` return on `ReturnUnionTypeRector`

**Comment by TomasVotruba** [comment]
Thank you :clap:
